flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -691,11 +716,15 @@ private ApplicationReport startAppMaster(
 							1));
 		}
 
-		final Set<File> userJarFiles = (jobGraph == null)
-				// not per-job submission
-				? Collections.emptySet()
-				// add user code jars from the provided JobGraph
-				: jobGraph.getUserJars().stream().map(f -> f.toUri()).map(File::new).collect(Collectors.toSet());
+		final Set<File> userJarFiles = new HashSet<>();
+		if (jobGraph != null) {
+			userJarFiles.addAll(jobGraph.getUserJars().stream().map(f -> f.toUri()).map(File::new).collect(Collectors.toSet()));
+		}
+
+		final List<URI> jarUrls = ConfigUtils.decodeListFromConfig(configuration, PipelineOptions.JARS, URI::create);
+		if (jarUrls != null && YarnApplicationClusterEntryPoint.class.getName().equals(yarnClusterEntrypoint)) {
```

Comment:
I am not sure why only for "application mode" we need to add pipeline jars to `userJarFiles`. Is is for python support?

Reply:
We also add for "per-job" mode in the if-branch above. That branch is just modified version of what exists currently on the master.

Reply:
You are right. For perjob mode, we also add the user jars to the `userJarFiles`.